services/tables/src/main/java/com/linkedin/openhouse/tables/controller/TablesController.java

```diff
@@ -476,4 +476,36 @@ public ResponseEntity<Void> purgeSoftDeletedTable(
     return new ResponseEntity<>(
         apiResponse.getResponseBody(), apiResponse.getHttpHeaders(), apiResponse.getHttpStatus());
   }
+
+  @Operation(
+      summary = "Restore Soft Deleted Table",
+      description =
+          "Restores a soft deleted table identified by tableId, databaseId, and deletedAtMs",
+      tags = {"Table"})
+  @ApiResponses(
+      value = {
+        @ApiResponse(responseCode = "204", description = "Soft Deleted Table RESTORE: NO CONTENT"),
+        @ApiResponse(responseCode = "400", description = "Soft Deleted Table RESTORE: BAD REQUEST"),
+        @ApiResponse(
+            responseCode = "401",
+            description = "Soft Deleted Table RESTORE: UNAUTHORIZED"),
+        @ApiResponse(responseCode = "403", description = "Soft Deleted Table RESTORE: FORBIDDEN"),
+        @ApiResponse(responseCode = "404", description = "Soft Deleted Table RESTORE: NOT FOUND"),
+        @ApiResponse(
+            responseCode = "500",
+            description = "Soft Deleted Table RESTORE: INTERNAL SERVER ERROR")
+      })
+  @PutMapping(value = "/v1/databases/{databaseId}/tables/{tableId}/restore")
+  public ResponseEntity<Void> restoreTable(
+      @Parameter(description = "Database ID") @PathVariable String databaseId,
+      @Parameter(description = "Table ID") @PathVariable String tableId,
+      @Parameter(description = "Deleted at timestamp in milliseconds")
+          @RequestParam(value = "deletedAtMs")
+          long deletedAtMs) {
+    com.linkedin.openhouse.common.api.spec.ApiResponse<Void> apiResponse =
+        tablesApiHandler.restoreTable(
```

Comment:
Should we have any check to ensure storage is not purged yet?

Reply:
Ideally when purging storage purging the soft deleted HTS reference would always come first. So if the table can still be restored then its soft deleted reference must still exist.